src/main/java/apoc/schema/Schemas.java

```diff
@@ -117,6 +135,101 @@ private AssertSchemaResult createUniqueConstraint(Schema schema, String lbl, Str
         return new AssertSchemaResult(lbl, key).unique().created();
     }
 
+    public List<AssertSchemaResult> recreateIndexes(long allIndexesTimeout, long dropAllIndexesTimeout, long recreateAllIndexesTimeout) throws ExecutionException, InterruptedException, IllegalArgumentException, TimeoutException {
+        Schema schema = db.schema();
+        List<AssertSchemaResult> result = Collections.synchronizedList(new ArrayList<>());
+        ExecutorService executorService = Pools.SINGLE;
+        List<IndexDefinition> indexesToRecreate = Util.inTxFuture(executorService, db, () -> {
+            List<IndexDefinition> idx = new ArrayList<>();
+            Iterable<IndexDefinition> outerIndexes = schema.getIndexes();
```

Comment:
not sure if we need to pull the schema from the db in the separate tx?

Reply:
does it have to be in a separate tx just reading them?

Reply:
It does! This line:

```
schema.getIndexes();
```

Takes a lock on all the indexes and it's only released when the transaction is closed. So you end up in a deadlock if you do that in the procedure's implicit transaction as it would still be holding the lock when the other thread is trying to drop it

Reply:
I meant I was not sure if in the separate tx we actually have to call `db.schema()...` there to have the right tx connected to the schema object.

Reply:
I'm not sure what you mean. The reason we have a separate thread to collect all the indexes and constraints is the following code:

![2019-11-15_13-25-00](https://user-images.githubusercontent.com/13220/68946652-685dc800-07ab-11ea-90cd-58b885fa372c.png)